### PR TITLE
tsp, support multi content types (generate protocol method only)

### PR DESCRIPTION
**Pull request by haolingdong-msft**
Link https://github.com/Azure/autorest.java/issues/1958

**Summary:**
In this pr, we add the logic to generate protocol methods only for multiple content types cases.

**Code changes:**
1. code-model-builder.ts: add the logic to disable generate convenient method when we found multiple content types are defined.
2. Add tests

**Test:**
1. multiple content types with `@sharedRoute`.
2. multiple content types with `@overload`.

### CODE REVIEW

**Comment by weidongxu-microsoft** [comment]
It is "multiple content types, regardless of overload or not".

File: typespec-extension/src/code-model-builder.ts
Comment (haolingdong-msft) [reply typespec-extension/src/code-model-builder.ts:543]: If regardless overload, below two cases will not generate convenient methods. But I think both case1 and case 2 should generate convenient method. Maybe there is some misunderstanding. To help us on the same page, could you explain what is the expected generated code for below two cases?
If regardless overload or not,  
Case 1:
```java
  @doc("one data type maps to multiple content types")
  @route("upload/single-body-type")
  @post
  uploadBytesWithSingleBodyTypeForMultiContentTypes(@body data: bytes, @header contentType: "application/octet-stream" | "image/jpeg" | "image/png"): void;
```
Case 2:
```java
@doc("multiple data types map to multiple content types using shared route")
  @route("upload/multi-body-types")
  @sharedRoute
  @post
  uploadBytesWithMultiBodyTypesForMultiContentTypes(@body data: bytes, @header contentType:  "application/octet-stream" | "image/jpeg" | "image/png"): void;

  @doc("multiple data types map to multiple content types using shared route")
  @route("upload/multi-body-types")
  @sharedRoute
  @post
  uploadJsonWithMultiBodyTypesForMultiContentTypes(@body data: Resource, @header contentType:  "application/json"): void;
```

Comment (weidongxu-microsoft) [reply typespec-extension/src/code-model-builder.ts:543]: case1 is multiple content-type, and should not generate convenience API.

same to uploadBytesWithMultiBodyTypesForMultiContentTypes in case2

uploadJsonWithMultiBodyTypesForMultiContentTypes in case2 is single content-type, which should have convenience API.

Comment (weidongxu-microsoft) [reply typespec-extension/src/code-model-builder.ts:543]: What is the point, if a badly written one like case1 generates convenience API, but if I write it better e.g. `@overload uploadData; @overload uploadPng; @overload uploadJpeg` now it won't generate convenience API?

Comment (weidongxu-microsoft) [reply typespec-extension/src/code-model-builder.ts:543]: Also, if you generate convenience API, I believe there will be a `ContentType` enum be 3 values.

What if that package get another one similar to case1, but just 2 content-type? Now you get another `ContentType` enum with 2 values?

**Comment by weidongxu-microsoft** [comment]
where is the "not generated overloaded API" part? e.g. uploadImagesWithOverload should not be generated at all.

**Comment by haolingdong-msft** [comment]
> where is the "not generated overloaded API" part? e.g. uploadImagesWithOverload should not be generated at all.

Sorry I'm not aware of this, do we have write-downs for this requirement?
To clarify the requirement, do we want to disable generation for protocol method and convenient method for all the operations with `@overload`?
For example, what will the generated code be like for op `uploadStringWithOverload`? 'no protocol+no convenient', or 'package-private protocol+no convenient'
```typescript
@doc("multiple data types map to multiple content types")
@route("upload/overload/multi-body-types")
@post
op uploadWithOverload(
  @body data: string | bytes | Resource,
  @header contentType: "text/plain" | "application/json" | "application/octet-stream" | "image/jpeg" | "image/png"
): void;

@doc("multiple data types map to multiple content types")
@route("upload/overload/multi-body-types")
@post
@overload(uploadWithOverload)
op uploadStringWithOverload(@body data: string, @header contentType: "text/plain"): void;
```

**Comment by weidongxu-microsoft** [comment]
It is "overload", so every method should be of name "uploadWithOverload" (maybe one with uploadWithOverloadWithResponse").

None of other name should appear at all, hence it be no protocol+no convenient, nothing, not even internal.

File: typespec-extension/src/code-model-builder.ts
Comment (weidongxu-microsoft) [inline typespec-extension/src/code-model-builder.ts:394]: Why not filter by `getOverloadedOperation` here?

If one can avoid `undefined` in type, avoid it.

Comment (haolingdong-msft) [reply typespec-extension/src/code-model-builder.ts:394]: good suggestion, will update. Thanks

**Comment by haolingdong-msft** [comment]
> ```ts
> uploadStringWithOverload
> ```

Update on 2023.5.24 sync up meeting with @srnagar, we shoud generate convenient method for operation `uploadStringWithOverload` named ``

> > where is the "not generated overloaded API" part? e.g. uploadImagesWithOverload should not be generated at all.
> 
> Sorry I'm not aware of this, do we have write-downs for this requirement? To clarify the requirement, do we want to disable generation for protocol method and convenient method for all the operations with `@overload`? For example, what will the generated code be like for op `uploadStringWithOverload`? 'no protocol+no convenient', or 'package-private protocol+no convenient'
> 
> ```ts
> @doc("multiple data types map to multiple content types")
> @route("upload/overload/multi-body-types")
> @post
> op uploadWithOverload(
>   @body data: string | bytes | Resource,
>   @header contentType: "text/plain" | "application/json" | "application/octet-stream" | "image/jpeg" | "image/png"
> ): void;
> 
> @doc("multiple data types map to multiple content types")
> @route("upload/overload/multi-body-types")
> @post
> @overload(uploadWithOverload)
> op uploadStringWithOverload(@body data: string, @header contentType: "text/plain"): void;
> ```


Update on 2023.5.24 sync up meeting with @srnagar, we shoud generate convenient method for operation `uploadStringWithOverload` named `uploadWithOverload`.

**Comment by haolingdong-msft** [comment]
Suggestion from @weidongxu-microsoft that it's better to support overload non-union method in seperate pr. So I will merge this pr first after approval and build pass.

File: typespec-extension/src/code-model-builder.ts
Comment (weidongxu-microsoft) [inline typespec-extension/src/code-model-builder.ts:556]: Do you need `toLowerCase()`?

Comment (haolingdong-msft) [reply typespec-extension/src/code-model-builder.ts:556]: https://github.com/microsoft/typespec/blob/main/packages/http/src/content-types.ts#L9 - it's case sensitive.

No matter using "contentType" or "ContentType" in tsp, the "parameter.name" here is `content-type`.

[playground](https://cadlplayground.z22.web.core.windows.net/cadl-azure/?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7CtIZcmVzdNUZdmVyc2lvbmluZ8wfYXp1cmUtdG9vbHMvyCstcHJvdmlkZXJodWLfLGMtxhVjb3Jl3yvIK3Jlc291cmNlLW1hbmFnZXIiOwoKdXNpbmcgVHlwZVNwZWMuSHR0cDvQFVJlc3TRFVbpANXIG0HEZS5Db3JlzhJSx3dNxnY7CgpAYXJtUOcA6k5hbWVzcGFjZQpAc2VydmljZSh7CiAgdGl0bGU6ICJDb250b3NvyC5IdWJDbGllbnQiLAogIOcBWTogIjIwMjEtMTAtMDEtcHJldmlldyIsCn0pCkBkb2MoyEUg6ACKIMhPIOYBCm1lbnQgQVBJLiIpCkB1c2VEZXBlbmRlbmN5KPUAw%2BgA%2FXMudjFfMF9QxnJfMSkKbugAziBNaWNyb3NvZnQu8gDD5AED5QCZbXVsdGlwbGUgZGF0YSDlAa8gbWFwIHRvIMkbY29udOQAoMUe5AChcm91dGUoInVwbG9hZC9vdmVyxQnFMS1ib2R5LcUrMcQscG9zdApvcCDGLVdpdGhPxzAxKAogIEDEL8V9OiBzdHLkAbR8IGJ5dGVzIHzpAS7kAWpAaGVh5AEw5ADMZW505AIGOiAgInRleHQvcGxhaW4iIHwgImFwcGxpY2F0aW9uL2pzb9IVb2N0ZXQtc3RyZWFtxR1pbWFnZS9qcGVnyw9wbmciCik6IHZvaWT%2FASv%2FASv%2FASvoASsy%2FgErMv8BK%2FYBK%2BcAjf8BK%2F8BK%2F8BK%2B8BKw%3D%3D)

Comment (weidongxu-microsoft) [reply typespec-extension/src/code-model-builder.ts:556]: Not "contentType" or "ContentType"

The `@header("Content-Type")`. They are different thing, please...

Comment (weidongxu-microsoft) [reply typespec-extension/src/code-model-builder.ts:556]:
```
@header("Content-Type") WhateverName:  "text/plain" | "application/json" | "application/octet-stream" | "image/jpeg" | "image/png"
```

Comment (haolingdong-msft) [reply typespec-extension/src/code-model-builder.ts:556]: Sorry, my fault. it's case insensitive, https://github.com/microsoft/typespec/blob/main/packages/http/src/content-types.ts#L13

Using `@header("Content-Type")` also works, and parameter.name is `Content-Type`. I will update the code.

File: typespec-tests/tsp/multi-content-types.tsp
Comment (weidongxu-microsoft) [inline typespec-tests/tsp/multi-content-types.tsp:71]: I assume typically for sharedRoute case, `uploadBytesWithMultiBodyTypesForMultiContentTypes` and `uploadJsonWithMultiBodyTypesForMultiContentTypes` should already cover the cases of 4 content-types, and hence dev won't write `uploadJsonOrBytesWithMultiBodyTypesForMultiContentTypes`.

**Review by weidongxu-microsoft — APPROVED** [review]

File: typespec-tests/tsp/multi-content-types.tsp
Comment (haolingdong-msft) [reply typespec-tests/tsp/multi-content-types.tsp:71]: Yes, just add this edge case to test our behavior because we don't have lint rule to forbid this.
I can remvoe this if you all feel it's unnecessary.

Comment (weidongxu-microsoft) [reply typespec-tests/tsp/multi-content-types.tsp:71]: Our test should be fine, even if you had it. Be careful on cadl-ranch one.

File: typespec-tests/tsp/multi-content-types.tsp
Comment (haolingdong-msft) [reply typespec-tests/tsp/multi-content-types.tsp:123]: no, it is not fixed, it still has compile error
[playground](https://cadlplayground.z22.web.core.windows.net/cadl-azure/?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7CtIZcmVzdNUZdmVyc2lvbmluZ8wfYXp1cmUtdG9vbHMvyCstcHJvdmlkZXJodWLfLGMtxhVjb3Jl3yvIK3Jlc291cmNlLW1hbmFnZXIiOwoKdXNpbmcgVHlwZVNwZWMuSHR0cDvQFVJlc3TRFVbpANXIG0HEZS5Db3JlzhJSx3dNxnY7CgpAYXJtUOcA6k5hbWVzcGFjZQpAc2VydmljZSh7CiAgdGl0bGU6ICJDb250b3NvyC5IdWJDbGllbnQiLAogIOcBWTogIjIwMjEtMTAtMDEtcHJldmlldyIsCn0pCkBkb2MoyEUg6ACKIMhPIOYBCm1lbnQgQVBJLiIpCkB1c2VEZXBlbmRlbmN5KPUAw%2BgA%2FXMudjFfMF9QxnJfMSkKbugAziBNaWNyb3NvZnQu8gDDOwoKCmludGVyZsQqVGVzdCDkAPQvLyBvdmVybG9hZAogIOYAu211bHRpcGxlIGRhdGEg5QHRIG1hcCB0byDJG2NvbnTkAMLFHiIpxDxyb3V0ZSgidXDETS%2FIVi%2FFMy1ib2R5LcstcG9zdAogIG9wIMYwV2l0aE%2FHMygKICAgIEDEM%2BUAgzogc3Ry5AHcfCBieXRlcyB86QFW5AGSICBAaGVh5AFa5wCV5AIwOiAidGV4dC9wbGFpbiIgfCAiYXBwbGljYXRpb24vanNv0hVvY3RldC1zdHJlYW3FHWltYWdlL2pwZWfLD3BuZyIKICApOiB2b2lkOwr%2FATb%2FATb%2FATb3ATZAyCco8gE9xCjpAVZT5QE4zB8o8gFXLP8BQGluIuoA7n0K)

File: typespec-extension/src/code-model-builder.ts
Comment (srnagar) [inline typespec-extension/src/code-model-builder.ts:563]:
```suggestion
    if (httpOperation.parameters.parameters
        && httpOperation.parameters.parameters.length > 0
        && httpOperation.parameters.parameters.some(
            (parameter) =>
                parameter?.type === "header" &&
                parameter?.name?.toLowerCase() === "content-type" &&
                parameter?.param?.type?.kind === "Union",
            )) {
        return true;
    }
    return false;
```

File: typespec-extension/src/code-model-builder.ts
Comment (srnagar) [inline typespec-extension/src/code-model-builder.ts:438]: It might be good to add additional details on why we are skipping this or link to a GitHub issue where we've discussed this.

File: typespec-extension/src/code-model-builder.ts
Comment (srnagar) [inline typespec-extension/src/code-model-builder.ts:470]: Same here - documenting the reason for skipping this will help understand the reason for this decision when someone is looking at this code in the future.

File: typespec-extension/src/code-model-builder.ts
Comment (srnagar) [inline typespec-extension/src/code-model-builder.ts:551]: Do we need to check for `length > 0`? `...parameters.some()` should handle empty list too right?

**Review by srnagar — APPROVED** [review]

**Pull request merged**
